sap/src/main/java/org/apache/zeppelin/sap/UniverseInterpreter.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zeppelin.sap;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.zeppelin.interpreter.Interpreter;
+import org.apache.zeppelin.interpreter.InterpreterContext;
+import org.apache.zeppelin.interpreter.InterpreterException;
+import org.apache.zeppelin.interpreter.InterpreterResult;
+import org.apache.zeppelin.interpreter.thrift.InterpreterCompletion;
+import org.apache.zeppelin.sap.universe.*;
+
+
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * SAP Universe interpreter for Zeppelin.
+ */
+public class UniverseInterpreter extends Interpreter {
+
+  public UniverseInterpreter(Properties properties) {
+    super(properties);
+  }
+
+  private UniverseClient client;
+  private UniverseUtil universeUtil;
+  private UniverseCompleter universeCompleter;
+
+  private static final String EMPTY_COLUMN_VALUE = StringUtils.EMPTY;
+  private static final char WHITESPACE = ' ';
+  private static final char NEWLINE = '\n';
+  private static final char TAB = '\t';
+  private static final String TABLE_MAGIC_TAG = "%table ";
+
+  @Override
+  public void open() throws InterpreterException {
+    String user = getProperty("universe.user");
+    String password = getProperty("universe.password");
+    String apiUrl = getProperty("universe.api.url");
+    String authType = getProperty("universe.authType");
+    this.client = new UniverseClient(user, password, apiUrl, authType);
+    this.universeUtil = new UniverseUtil();
+  }
+
+  @Override
+  public void close() throws InterpreterException {
+    try {
+      client.close();
+    } catch (Exception e) {
+      throw new InterpreterException(e.getCause());
+    }
+  }
+
+  @Override
+  public InterpreterResult interpret(String st, InterpreterContext context)
+      throws InterpreterException {
+    try {
+      InterpreterResult interpreterResult = new InterpreterResult(InterpreterResult.Code.SUCCESS);
+      String paragraphId = context.getParagraphId();
+      String token = client.getToken(paragraphId);
+      client.loadUniverses(token);
```

Comment:
Does client need to load token for each interpret call ? If not, then you can load token in open()

Reply:
Yes, because user has limit of sessions and session has TTL. If the session expires it hangs.